Add storage factories to easily swap out vector db implementations

smenon8 opened the pull request.

thompsonmj reviewed (changes requested):
Excellent architecting!

I think the tests need updated as well though to use new storage module. I think there might be issues with relative imports as well. But for the most part, I think this looks good and can be debugged further if needed after merging. Double check the chroma storage interface name assignment for cleanup in case I misunderstood something.

Next step should be packaging!

smenon8 commented on src/bioclip_vector_db/vector_db.py:
```diff
@@ -152,7 +135,7 @@ def _load_database_web(self):
                 continue
 
-            self._collection.add(embeddings=[embedding], ids=[id])
+            self._storage.add_embedding(embeddings=[embedding], ids=[id])
```
Sorry, this was me being lousy. Fixed.

smenon8 commented on src/bioclip_vector_db/storage/storage_impl.py:
```diff
@@ -0,0 +1,59 @@
+        if "collection_dir" not in kwargs:
+            raise "Chromadb cannot be initialized without collection_dir."
+
+        self._metadata = kwargs["metadata"]
```
reset() can just use _collection.name

smenon8 commented:
> Excellent architecting!
> 
> I think the tests need updated as well though to use new storage module. I think there might be issues with relative imports as well. But for the most part, I think this looks good and can be debugged further if needed after merging. Double check the chroma storage interface name assignment for cleanup in case I misunderstood something.
> 
> Next step should be packaging!

Thanks for such detailed review. I have addressed your comments and fixed the test (the test needs a lot more work!!). 

I will send out a quick PR tomorrow for packaging this.
